pkg/ingester/limiting.go

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/VividCortex/ewma"
+	"github.com/go-kit/log/level"
+	"github.com/shirou/gopsutil/v3/process"
+	"github.com/weaveworks/common/httpgrpc"
+)
+
+func (i *Ingester) updateResourceUtilization() {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return
+	}
+
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+
+	proc, err := process.NewProcess(int32(os.Getpid()))
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get process info",
+			"err", err.Error())
+		return
+	}
+
+	memInfo, err := proc.MemoryInfo()
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get memory info",
+			"err", err.Error())
+		return
+	}
+	i.memoryUtilization.Store(memInfo.RSS)
+
+	cput, err := proc.Times()
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get process times",
+			"err", err.Error())
+		return
+	}
+
+	now := time.Now().UTC()
+
+	lastCPUTime := i.lastCPUTime.Load()
+	// TODO: Should Guest and GuestNice be ignored, as potentially duplicated in User?
+	// https://github.com/mackerelio/go-osstat/blob/54b9216143f9aa087151258f103dbd5c381f7915/cpu/cpu_linux.go#L70-L74
+	cpuTime := cput.User + cput.System + cput.Idle + cput.Nice + cput.Iowait + cput.Irq +
+		cput.Softirq + cput.Steal + cput.Guest + cput.GuestNice
+	i.lastCPUTime.Store(cpuTime)
+
+	i.lastResourceUtilizationUpdate.Store(now)
+
+	if lastUpdate.IsZero() {
+		return
+	}
+
+	cpuUtil := (cpuTime - lastCPUTime) / now.Sub(lastUpdate).Seconds()
+	// Shift samples to the left, making room for another
+	for j := 1; j < len(i.cpuLoadSamples); j++ {
+		i.cpuLoadSamples[j-1] = i.cpuLoadSamples[j]
+	}
+	i.cpuLoadSamples[len(i.cpuLoadSamples)-1] = cpuUtil
+	a := ewma.NewMovingAverage(float64(len(i.cpuLoadSamples)))
+	for _, s := range i.cpuLoadSamples {
+		a.Add(s)
+	}
+
+	cpuA := a.Value()
+	i.cpuUtilization.Store(cpuA)
```

Comment:
@jhalterman WDYT of my usage of EWMA here (window of 60 seconds)?

Reply:
The way the EWMA is meant to be used is that we add samples to the same EWMA instance over time. So here, I don't think we need to store `cpuLoadSamples` separately, we'd just add the latest sample to our EWMA and get whatever the updated `.Value()` is.

Reply:
@jhalterman ah yep, you're right. The library's docs made me think as if it would maintain a window (added to with `.Add`), but in fact it just keeps a scalar. Will fix.

Reply:
Should now be fixed.